sale_unreconciled/models/sale_order.py

```diff
@@ -0,0 +1,142 @@
+
+    def action_view_unreconciled(self):
+        self.ensure_one()
+        acc_item = self.env["account.move.line"]
+        domain = self._get_sale_unreconciled_base_domain()
+        unreconciled_domain = expression.AND(
+            [domain, [("sale_order_id", "=", self.id)]]
+        )
+        unreconciled_items = acc_item.search(unreconciled_domain)
+        action = self.env.ref("account.action_account_moves_all")
+        action_dict = action.read()[0]
+        action_dict["domain"] = [("id", "in", unreconciled_items.ids)]
+        return action_dict
+
+    def action_reconcile(self):
+        self.ensure_one()
+        acc_item = self.env["account.move.line"]
+        domain = self._get_sale_unreconciled_base_domain()
+        unreconciled_domain = expression.AND(
+            [domain, [("sale_order_id", "=", self.id)]]
+        )
+        unreconciled_items = acc_item.search(unreconciled_domain)
+        writeoff_to_reconcile = False
+        for account in unreconciled_items.mapped("account_id"):
+            acc_unrec_items = unreconciled_items.filtered(
+                lambda ml: ml.account_id == account
+            )
+            all_aml_share_same_currency = all(
+                [x.currency_id == self[0].currency_id for x in acc_unrec_items]
+            )
+            writeoff_vals = {
+                "account_id": self.company_id.sale_reconcile_account_id.id,
+                "journal_id": self.company_id.sale_reconcile_journal_id.id,
+                "sale_order_id": self.id,
+                "currency_id": self.currency_id.id,
+            }
+            if not all_aml_share_same_currency:
+                writeoff_vals["amount_currency"] = False
+            if writeoff_to_reconcile:
+                writeoff_to_reconcile += unreconciled_items._create_so_writeoff(
+                    writeoff_vals
+                )
+            else:
+                writeoff_to_reconcile = unreconciled_items._create_so_writeoff(
+                    writeoff_vals
+                )
+        # add writeoff line to reconcile algorithm and finish the reconciliation
+        if writeoff_to_reconcile:
+            remaining_moves = unreconciled_items + writeoff_to_reconcile
+        else:
+            remaining_moves = unreconciled_items
+        # Check if reconciliation is total or needs an exchange rate entry to be created
+        if remaining_moves:
+            remaining_moves.filtered(lambda l: not l.reconciled).reconcile()
+            return {
+                "name": _("Reconciled journal items"),
+                "type": "ir.actions.act_window",
+                "view_type": "form",
+                "view_mode": "tree,form",
+                "res_model": "account.move.line",
+                "domain": [
+                    ("id", "in", unreconciled_items.ids + writeoff_to_reconcile.ids)
+                ],
+            }
+
+    def action_done(self):
+        if self.unreconciled:
```

Comment:
this should be refactored to iterate over self, as in the original implementation there is no self.ensure_one ()

Reply:
good point! done :)